쥬스 메이커 [STEP 1] bradheo65, bonf, ZZBAE

ZZBAE opened the pull request:
안녕하세요 GREEN! @GREENOVER 
이번 쥬스메이커 프로젝트를 함께 하게 된 bradheo65, bonf, ZZBAE 입니다.
step 1 PR 잘 부탁드리겠습니다!

## 목차
1. 순서도 작성
2. UML 작성
3. 고민한 점 + 해결한 점
4. 조언을 구하고 싶은 부분

## 1. 순서도 작성
![unknown](https://user-images.githubusercontent.com/98302604/165425608-3a918bc5-32e5-4f12-b551-302fe8529ff9.png)

> 선택한 과일 쥬스에 대해 재고확인 후 해당 과일 쥬스에 대한 재료를 소모 해서 프로그램을 끝내거나
해당 과일 재료가 없다면 에러를 발생시킬 수 있게 만들어보았습니다.

## 2. UML 작성
![unknown-2](https://user-images.githubusercontent.com/98302604/165425706-b232ecd2-79ed-4c5f-8064-06484b0cfeef.png)


> 프로젝트 시작 전 Green이랑 대화를 나눴을 때, UML이라는 것을 처음 알게 되었습니다.
>  참고하라고 알려주신 링크 " [클래스 다이어그램 작성법](https://gmlwjd9405.github.io/2018/07/04/class-diagram.html) " 를 읽고 다같이 많은 시간을 할애하여 작성해보았습니다. 프로젝트를 하면서 필요한 함수들의 정리, 각 타입에 대한 함수들의 기능, 객체 정의 등 다같이 얘기하면서 정리하는 시간을 갖고 나니 코드를 작성하는데 큰 도움이 되었습니다.
>  사실 작성하면서 정확하게 썼는지는 의문이긴 하지만, 수정이 필요하거나 추가하면 좋을 것 같은 부분들을 알려주시면 앞으로 UML에 대한 공부하는데 큰 도움이 될 것 같습니다 ☺️

## 3. 고민한 점 / 해결한 점
#### 1) 열거형 프로토콜
---
- ***고민한점:*** 과일 재고의 초기값을 설정하기 위해 처음에 딕셔너리 형태로 하드코딩을 진행보았지만 다시 생각해보니 이미 열거형을 사용해서 초기값을 넣어준다면 유지보수하기도 편하기 때문에 어떻게 구현을 하면 될지 고민해보았습니다.

- ***해결한점:*** Fruits라는 열거형에 CaseIterable이라는 프로토콜을 설정해주어서 딕셔너리 형태로 사용할 수 있도록 설정해 주었습니다.

```swift
enum Fruits: CaseIterable {
    case strawberry
    case banana
    case pineapple
    case kiwi
    case mango
}
```
#### 2) for문과 forEach의 사용성
---
- ***고민한점:*** for문과 forEach 중 어떤 것을 사용해서 각 과일 별로 초기 값을 넣어줄지 고민했습니다.

- ***해결한점:*** 서로의 차이점은 for문인 경우 지정된 요소 수 만큼 반복되고 forEach인 경우 저장된 요소 수 만큼 반복되어 전체에 하나씩 넣어 줄 수 있어 forEach를 사용해서 작성해 보았습니다.
> 
```swift
    private var stocks: [Fruits: Int] = [:]
    private let stock = 10
    init() {
        Fruits.allCases.forEach { stocks[$0] = stock }
    }
```
#### 3) try-catch 디폴트 오류
---
- ***고민한점:*** try-catch 문을 이용하여 오류 처리를 하는 과정에서 "call can throw, but it is not marked with 'try' and the error is not handled" 문구의 오류가 발생하였습니다. 

- ***해결한점:*** try-catch 오류의 디폴트값을 설정해서 해결할 수 있었습니다.
```swift
    do {
        try fruitStore.chooseRecipe(order: juice)
    } catch StockError.outOfError {
        print("재고부족")
    } catch { // 디폴트 오류를 설정
        print(error)
    }
```


## 4. 조언을 구하고 싶은 부분

- Green이 말씀해 주셨듯이 예를 들어, 10시간 중에 9시간을 UML 작성에 투자하고 1시간만에 코드를 짤 수 있게끔 하라고 추천해주신게 인상깊었습니다. 앞으로 UML을 작성할 때, 가장 중요하게 생각해야하는 포인트가 어떤게 있을지 궁금합니다! 👩🏻‍💻

- JuiceMaker는 struct로, FruitStore는 Class로 생성된 이유에 대해서 고민해 보았습니다. JuiceMaker는 아르바이트생이 바뀌듯 인스턴스가 바뀌어야 하기 때문에 struct로 생성했다고 생각했습니다. FruitStore에 저장된 프로퍼티(과일 개수)는 항상 일정해야 되기 때문에 참조 형식인 Class로 생성해야 된다고 생각했습니다. 다른 이유가 더 있을지 혹은 조언해 주실 부분에 대해서 알고 싶습니다! 😆

bradheo65 committed:
feat: 쥬스메이커와 FruitStore 열거형 추가, 초기 값, 쥬스 종류 별 필요한 재료 함수 작성

ZZBAE committed:
feat: Fruit 삭제 및 Juice 생성, choiceJuice case 정리

apwierk2451 committed:
feat: 과일 개수 초기값 설정, 주문 받은 것 처리 하는 함수 추가
1. 주문 받고 오류처리하는 함수
2. 쥬스에 필요한 과일 개수 처리하는 함수

bradheo65 committed:
style: 재고 관리함수명, 쥬스 관리 변수명 변경

ZZBAE committed:
style: 주스를 주문하는 함수와 주스 재료인 과일의 변수 이름 변경,  파일 별 주석 수정 및 필요없는 주석 제거

ZZBAE committed:
Merge branch 'feature/step01' of https://github.com/ZZBAE/ios-juice-maker into feature/step01

# Conflicts:
#	JuiceMaker/JuiceMaker/Model/FruitStore.swift
#	JuiceMaker/JuiceMaker/Model/Juice.swift

GREENOVER commented on JuiceMaker/JuiceMaker.xcodeproj/project.pbxproj:
```diff
@@ -18,6 +21,9 @@
 /* End PBXBuildFile section */
 
 /* Begin PBXFileReference section */
+		18A346C328181BBA003F2F29 /* Fruits.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = Fruits.swift; sourceTree = "<group>"; };
```
Model 그룹안에 Fruit, FruitStore, Juice, JuiceMaker를 관심사있게 잘 넣어주신점 아주 좋아요👍
그런데 혹시 모델 그룹안에서도 관심사별로 조금 더 그룹핑을 해볼 수 있을거 같고 그렇다면 더 가독이 좋아질것 같은데 어떨까요?🙋🏻
예를들어 Fruit와 FruitStore는 Fruit라는 그룹으로 Juice와 JuiceMaker는 Juice라는 그룹으로 묶을 수 있을것 같습니다!
지금이야 파일이 별로 없어 모델 그룹안에 여러 파일이 있어도 혼동이 없지만 추후 무수히 많아지게 될 경우를 대비해 그룹핑을 해놓으면 좋지 않을까 의견 드립니다🙌

bradheo65 replied:
그린의 의견을 반영해서 Fruit 그룹안에 Fruit, FruitStore, Juice 그룹 안 Juice, JuiceMaker 그룹으로 묶어보았습니다!

GREENOVER commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -11,8 +11,8 @@ class ViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
         // Do any additional setup after loading the view.
+        let a = JuiceMaker()
```
만약 현재 사용되지 않는 단순 테스트 용도의 코드라면 PR시 제거하는 방향이 맞을것 같습니다🙋🏻

bradheo65 replied:
네! 미쳐 생각을 못했네요.😅

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+enum Fruit: String {
+    case strawberry = "딸기"
+    case banana = "바나나"
+    case pineapple = "파인애플"
+    case kiwi = "키위"
+    case mango = "망고"
```
rawValue로 활용하신 이유가 궁금합니다🙋🏻
nested type과의 차이가 무엇일까요?
(현재 방식이 잘못되었다는 얘기는 전혀 아닙니다!👍👍)

bradheo65 replied:
앗 특정한 이유가 있어서 사용한 건 아닙니다!! 단지 코드 가독성 떄문에 기입 했었습니다.

GREENOVER replied:
아하~ 그럼 어떤 둘은 어떤 차이가 있으며 현재와 추후를 생각했을때 어떤 방식이 더 적합할거라 같다고 생각하실까요~?

apwierk2451 replied:
nested type을 이용할 경우 Fruit타입으로 저장이 되는 걸로 알고 있습니다.
rawValue를 이용할 경우 String타입으로 나오므로 나중에 elert메세지 등 에 사용할 수도 있겠다는 생각을 했습니다.

GREENOVER replied:
명확한 이유를 가져서 좋네요..! 
그럼 만약에 현재는 과일에 대해 String 값만 가지게 되지만 딸기라는 케이스에서 딸기라는 String값과 수량이라는 Int의 값을 가지고 이를 원할때 필요한걸 써줘야한다면 어떻게 해줘야할까요?🙋‍♂️

ZZBAE replied:
Fruit을 key값으로 갖는 Dictionary를 새롭게 하나 더 만들어서 Int값을 전해주는 식으로 했는데, 혹시 더 좋은 방법이 있을까요?

GREENOVER replied:
정말 궁금해서 여쭤봤던 부분인데 볼륨이 커졌군요!ㅋㅋ
저는 요런식도 있다고 생각했어요..!
```swift
enum Fruit: CaseIterable {
    case strawberry
    case banana
    case pineapple
    case kiwi
    case mango
    
    var name: String {
        switch self {
        case .strawberry:
            return "딸기"
        case .banana:
            return "바나나"
        case .pineapple:
            return "파인애플"
        case .kiwi:
            return "키위"
        case .mango:
            return "망고"
        }
    }
    
    var count: Int {
        switch self {
        case .strawberry: 
            return 10
       ....
    ....
        }
    }
}
```
이런 연산 프로퍼티로 이용한다면 조금 더 명확하고 편리하지 않을까 생각했습니다!
지금 구현에서는 당장 필요없고 rawValue로 충분하겠지만요🙌

GREENOVER commented on JuiceMaker/JuiceMaker.xcodeproj/project.pbxproj:
```diff
@@ -18,6 +21,9 @@
 /* End PBXBuildFile section */
 
 /* Begin PBXFileReference section */
+		18A346C328181BBA003F2F29 /* Fruits.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = Fruits.swift; sourceTree = "<group>"; };
```
Fruit.swift는 디렉토리에 만들어져있긴한데 Xcode 빌드 파일 디렉토리에는 없네요..?🥲

bradheo65 replied:
브랜치를 나누어서 작업 후에 merge 하는 과정에서 디렉토리가 변경된 것 같습니다.

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+enum Fruit: String {
```
해당 Fruit.swift 파일에 궁금한점이 있습니다!
현재 Xcode에서는 디렉토리에 Fruit 파일이 없으며 해당 파일은 JuiceMaker 디렉토리에 파일로만 들어가있습니다!
해당 현상이 일어난데 있어 두가지 추측되는 부분이 있습니다.
1. 아마 Xcode에서 생성한것이 아닌 파일을 단순히 해당 디렉토리에 넣었기에 Xcode proj 파일내에서 인덱싱을 못함
2. 머지 혹은 컨플릭 처리 중 디렉토리가 변경되었거나 해당 파일에 대한 디렉토리 연결 코드가 사라짐
혹시 이 해당 Fruit 파일을 사용하시는것이면 Xcode에서 인식할 수 있게 해보시면 어떨까 합니다🙋🏻
(필요없다면 과감히 해당 파일 삭제를..!)

bradheo65 replied:
브랜치를 나누어서 작업 후에 merge 하는 과정에서 디렉토리가 변경된 것 같습니다.

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
-// 과일 저장소 타입
 class FruitStore {
     
+    private var stocks: [Fruits: Int] = [:]
```
딕셔너리 사용 아주 아주 좋네요👍👍

bradheo65 replied:
감사합니다!🫡

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
-// 과일 저장소 타입
 class FruitStore {
     
+    private var stocks: [Fruits: Int] = [:]
+    private let stock = 10
```
약간 두 프로퍼티가 혼동됩니다🥲
stocks은 과일과 재고수량을 담은 stock들인걸로 해석이 되는데
그렇다면 아래 단순 수량을 나타내는 stock은 무언가 Fruit: Int의 딕셔너리 타입으로 생각되는 흐름인데 실제적으로는 정수타입이라 그런것 같아요..!
네이밍을 조금 더 명확히 해보면 어떨까요~?🙋🏻

GREENOVER replied:
stock (수량)에 대한 값을 요렇게 FruitStore 클래스안에서 선언 정의해주는것과 이니셜라이저로 받아와 값을 주어주는것은 어떤 차이가 있을까요?
어떤게 앞으로 유지보수 혹은 코드가 확장되었을때 유리할까요?🤔

bradheo65 replied:
stock -> defaultFruitStock 변경했습니다.

Juice에서 FruitStore을 불러올떄 이니셜라이저로 초기값을 설정해주는 방법으로 변경해보았습니다.
이 방법말고 다른 방법에 대해서도 한번 생각해보겠습니다.
```swift
struct JuiceMaker {
    let fruitStore = FruitStore(defaultFruitStock: 10)
}
```

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+    private var stocks: [Fruits: Int] = [:]
+    private let stock = 10
+    init() {
+        Fruits.allCases.forEach { stocks[$0] = stock }
```
크.. 고급지게 이니셜을 해주셨네요👍
Fruits 타입을 CaseIterable 프로토콜 채택으로 해당 고급진 표현이 가능했겠네요!
forEach 사용도 아주 나이스하네요👍👍

bradheo65 replied:
🫡

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+        }
+    }
+
+    private func checkFruitStock(juice: Juice, fruits: Fruits) throws {
```
접근제어 레벨 아주 잘 구분해서 명확하게 사용해주셨네요👍

bradheo65 replied:
🫡

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+        Fruits.allCases.forEach { stocks[$0] = stock }
+    }
+    
+    func chooseRecipe(order: Juice) throws {
```
메서드 네이밍만 본다면 FruitStore가 해주는 역할은 아닌것 같아요..!
"레시피를 선택하다"라는 메서드명으로 해석되는데 해당 메서드는 느낌적으로 JuiceMaker가 해줄것 같다고 느껴졌어요🤔
결국 FruitStore에서 해주는 역할은 세가지로 나눠볼 수 있을것 같은데요.
1. 쥬스를 만들기 위해 요구되는 과일재고가 존재하는지 검증 (check)
2. 과일재고 사용 (use)
3. 과일재고 업데이트 (update)
로 정의해볼 수 있을것 같은데요.
그렇다면 해당 메서드는 useStock과 같은 네이밍이 어울릴것 같습니다🤔

ZZBAE replied:
chooseRecipe = "레시피를 선택하다"라는 메서드명에서, 추천해주신 2. 과일재고 사용 (use) 를 참고해서 조금 더 디테일하게
"useFromFruitStock" 라고 바꾸어보았습니다. 혹시 쓸데없이 디테일하다거나 너무 길다거나 하진 않을까요?🧐

GREENOVER replied:
아뇨 좋습니다!
네이밍이 디테일하고 길다고 올바르지 않은 네이밍은 아니라고 생각해요 저는.
오히려 더 길더라도 명확하게 해당 메서드명만 보고서 어떤 메서드구나~ 어떤 역할을 하겠구나~ 유추할 수 있는 네이밍이 올바른 방향이라고 생각합니다ㅎㅎ
하나 궁금한게 있습니다!
useFromFruitStock과 useFruitStock은 의미상으로 큰 차이가 있을까요?
From이라는 전치사를 사용해주신 의도가 궁금합니다🙋🏻
(제가 영알못이라ㅎ..)

ZZBAE replied:
저는 개인적으로 useFruitStock은 과일재고 전체를 사용하는 느낌이여서 전치사 From을 붙여서 과일재고 에서 필요한걸 사용한다는 느낌을 내려고 더 디테일하게 작성을 했던  것 같습니다..!

GREENOVER replied:
오오 그럴 수 있겠네요!👍👍👍

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+    func chooseRecipe(order: Juice) throws {
+        for fruit in order.chooseJuice.keys {
+            try checkFruitStock(juice: order, fruits: fruit)
+        }
```
추가로 checkFruitStock 즉 과일재고수량 검증하는것만 단순히 호출하고 해당 메서드는 다른 역할을 해주는것 같지 않습니다🤔
그렇다면 아래 makeJuice의 구현과 합쳐보면 어떨까요?
로직적으로 생각해보면 checkFruitStock을 호출할거고 그 뒤로 makeJuice 내에 있던 stocks을 업데이트 쳐주는것이 하나의 해당 메서드에서 구현될 수 있을것 같습니다🙋🏻

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+        guard let stockFruit = stocks[fruits] ,let needFruitAmount = juice.chooseJuice[fruits], stockFruit >= needFruitAmount else {
+            throw StockError.outOfError
+        }
+        makeJuice(juice: juice, fruits: fruits)
```
해당 메서드를 여기서 호출하는게 어색하다고 느껴졌습니다🥲
그 이유는 checkFruitStock이라는 메서드 시그니쳐만 본다면 단순히 과일재고 검증에 대한 메서드일테고 해당 메서드를 호출할때도 "과일 재고에 대해 검증해줘~" 라는 명령을 내렸는데 과일재고를 검증하고 만들어버리기까지 해버리는 느낌이여서 객체지향적으로 조금 어색해보여요🥲

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+    }
+
+    private func checkFruitStock(juice: Juice, fruits: Fruits) throws {
+        guard let stockFruit = stocks[fruits] ,let needFruitAmount = juice.chooseJuice[fruits], stockFruit >= needFruitAmount else {
+            throw StockError.outOfError
+        }
```
한줄로 쭈욱 이어지는 코드 컨벤션보다 이런 컨벤션은 어떠신가요?
```swift
guard let stockFruit = stocks[fruits],
      let needFruitAmount = juice.chooseJuice[fruits],
      stockFruit >= needFruitAmount
else {
    throw StockError.outOfError
}
```
사실 실제적으로 요러한 컨벤션을 많이 사용하긴 합니다!
저는 그 이유가 개발자들은 가로로 긴것보다 세로로 긴걸 더 선호한다고 생각해요.
장비마다 모두 가로 사이즈는 다를텐데 작은 13인치의 장비라면 이렇게 길어질 경우 가독성이 떨어지게 되더라구요.
그래서 다른 많은 레퍼런스를 참고해보면 위와 같은 개행 컨벤션을 가져가는걸 볼 수 있습니다🙋🏻

bradheo65 replied:
한번 생각해보니 그렇네요..😓 좋은 의견 감사합니다!!😃

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+    }
+
+    private func checkFruitStock(juice: Juice, fruits: Fruits) throws {
+        guard let stockFruit = stocks[fruits] ,let needFruitAmount = juice.chooseJuice[fruits], stockFruit >= needFruitAmount else {
```
컴파일에 전혀 문제는 안되지만 stocks[fruits] 뒤에 comma의 위치가 조금 어색합니다!
단순 오타일것 같습니다🙋🏻

ZZBAE replied:
넵 오타 수정했습니다!

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+        makeJuice(juice: juice, fruits: fruits)
+    }
+    
+    private func makeJuice(juice: Juice, fruits: Fruits) {
```
이 친구도 네이밍이 어색합니다🥲
쥬스를 만드는것은 JuiceMaker가 하는 역할이지 FruitStore 객체가 해주는 역할은 아닐것 같아요🤔
실제적으로 구현만 본다면 useStock 혹은 updateStock, changeStock과 같은 네이밍이 어울릴것 같습니다🙋🏻

ZZBAE replied:
FruitStore 객체가 해주는 역할을 고려하여 팀원들과의 대화 끝에 그린이 추천해주신 useStock을 사용하게 되었습니다!

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+    }
+    
+    private func makeJuice(juice: Juice, fruits: Fruits) {
+        if let stockFruit = stocks[fruits] ,let requiredFruit = juice.chooseJuice[fruits] {
```
요 부분도 적절한 개행과 stock[fruits] 뒤에 comma가 한칸 앞으로 오면 더 가독이 좋을것 같습니다🙋🏻

ZZBAE replied:
수정 완료하였습니다! 🫡

GREENOVER commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,34 @@
+
+    private func checkFruitStock(juice: Juice, fruits: Fruits) throws {
+        guard let stockFruit = stocks[fruits] ,let needFruitAmount = juice.chooseJuice[fruits], stockFruit >= needFruitAmount else {
+            throw StockError.outOfError
```
어떤 과일 재고가 부족한지에 대한 정보는 추후 필요없을까요?🙋🏻

ZZBAE replied:
해당 기능에 대해 대화를 해보았고, 앞서 말씀드린 것 처럼 새로운 Dictionary를 만들어서 부족한 재고에 대한 정보를 만들어진 Dictionary에 넣어서 오류처리하는 기능을 만들어보려고 하고 있습니다! 🫡

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Fruits.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by bradheo65, bonf, ZZBAE on 2022/04/26.
+//
+
+enum Fruits: CaseIterable {
```
CaseIterable 사용 아주 좋네요👍👍

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Fruits.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by bradheo65, bonf, ZZBAE on 2022/04/26.
+//
+
+enum Fruits: CaseIterable {
```
enum 타입의 네이밍에서는 대부분 단수형을 사용합니다!
과일들의 종류보다 과일의 종류가 더 어색하지 않은것처럼요😀
(Juice 타입의 Enum을 네이밍 해주신것처럼요!)

bradheo65 replied:
넵 알겠습니다! 수정하겠습니다.

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,35 @@
+    case kiwi
+    case pineapple
+
+    var chooseJuice: [Fruits: Int] {
```
프로퍼티 네이밍으로 "쥬스를 선택하다"가 되는게 어색해보입니다🥲
실제적으로 구현을 보면 recipe라는 네이밍은 어떨까요?
실제적으로 해당 프로퍼티를 호출하는 시그니쳐를 보면 `strawberry.chooseJuice`인데
직역을 해보자면 `딸기의 쥬스선택`인 느낌이라 보여집니다..!

ZZBAE replied:
딸기의 쥬스선택이라니.. 영어의 사용이 미숙하여 발생된 문제인 것 같습니다ㅎㅎ.. 😖
추천해주신대로 chooseJuice -> recipeOfJuice 로 변경했습니다 !

GREENOVER commented on JuiceMaker/JuiceMaker/Model/StockError.swift:
```diff
@@ -0,0 +1,10 @@
+//
+
+enum StockError: Error {
+    case outOfError
```
에러명이 조금 어색합니다🥲
"에러 부족" 느낌인데 재고가 없다는 에러라면 outOfStock과 같은 네이밍을 유도하신건지 궁금합니다🙋🏻
추가로, 해당 에러외에 stock에서 발생시킬 수 있는 에러는 더 없을까요?🙋🏻

ZZBAE replied:
outOfError 가 아닌 outOfStock 인데, 실수한 것 같습니다..! 🥲 

그리고 지금 당장은 재고가 부족할 때 생기는 에러 외에 다른 에러가 발생할 경우를 생각해내지 못하고 있습니다,, 혹시 저희가 놓치고 있는 부분이 있을까요?

GREENOVER replied:
현재 스텝에서 구현해주신 바에서는 없을 수 있습니다!
다만 에러처리에 있어 어떤 에러가 방출될지 모르는 경우도 존재할 수 있다고 생각해요.
그래서 저는 알 수 없는 에러라는 측면에서 unknown 에러 케이스를 하나 추가해줘서 사용해주곤 합니다!
사실 모든 에러를 다 초반에 설계하고 파악하기에 어려우니 요렇게 default같은 느낌의 에러를 만들어준다고 보면 좋을것 같아요🙋🏻

ZZBAE replied:
알 수 없는 에러 케이스를 추가해준다는건 생각지 못한 부분인데, 알려주셔서 감사합니다! 이후의 step들에서 발생될 오류에 대해서 추가로 얘기를 해보겠습니다. 🥸

GREENOVER commented on JuiceMaker/JuiceMaker/Model/StockError.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by bradheo65, bonf, ZZBAE on 2022/04/26.
+//
+
+enum StockError: Error {
```
해당 에러 타입을 모델 그룹에 위치시키는것보다 별도의 Error 그룹을 만들어 관리해보면 어떨까요?
과일, 쥬스 모델 타입들과 관심사가 다르다고 생각합니다..!
추후 더 많은 에러 타입들과 파일이 생길것 같습니다🙋🏻

ZZBAE replied:
수정 완료하였습니다! 🫡

GREENOVER commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,19 @@
 struct JuiceMaker {
+    let fruitStore = FruitStore()
     
+    func orderFruitJuice(of juice: Juice){
```
쥬스메이커가 "과일쥬스를 주문한다" 라는 것은 조금 어색해보입니다🤔
FruitStore에 위치하고 있던 makeJuice라는 네이밍이 여기오는게 더 적합할것 같다고 느껴지는데
여러분 생각은 어떠신가요~?🙋🏻

ZZBAE replied:
말씀해주신대로 주문받은 주스를 만드는 기능이 makeJuice라는 네이밍에 더 부합하는 것 같아서 변경하였습니다! 😊

GREENOVER commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,19 @@
+    let fruitStore = FruitStore()
     
+    func orderFruitJuice(of juice: Juice){
+        do {
+            try fruitStore.chooseRecipe(order: juice)
+        } catch StockError.outOfError {
+            print("재고부족")
+        } catch {
+            print(error)
+        }
```
에러 처리를 아주 잘 해주셨네요👍
추가로 생각해보면 좋을 부분이 두가지 있습니다..!
1. catch를 해주는 부분 우선 아주 좋고,
추후 catch 시 다른 구현이 올테지만 현재 StockError.outOfError를 캐치하는걸 봐보면,
FruitStore에서 과일재고에 대해 부족하다면 throw로 던져줘 여기서 처리를 해주는걸로 이해했습니다.
그러면 예를들어 망고키위 쥬스에서 망고/키위 재고가 모두 부족하다면 해당 catch에서 두개를 각자 throw한걸 받을텐데 추후 UI에 "망고/키위 모두 부족해요!"와 같이 인지할 수 있을까요?🤔
2. PR 메시지에도 작성해주셨던것 같은데 현재 StockError enum을 가져오니 만약 해당 케이스에 없는 즉 디폴트 에러에 대해서는 15~16과 같이 잘 처리해주셨다고 보입니다!
그런데 단순 error를 프린트해주는것보다 현재와 같은 경우는 "알 수 없는 에러"와 같은 맥락으로 보입니다!
그렇기에 Error를 정의해준 부분에서 각자 에러에 대해 케이스별 String 타입의 description을 만들어 14와 16번째 라인에서 print문에 해당 description을 사용하면 어떨까 합니다!🙋🏻

apwierk2451 replied:
그린의 1번 리뷰에 대해서 생각해 본 결과 재고가 부족한 경우는 UI에서 값을 이용할 수 있도록 오류처리하지 않도록 하려고 합니다. 제가 잘 이해한 것이 맞을까요?

GREENOVER replied:
네! 아주 잘 이해하신것 같습니다👍
혹시 모르니 제가 조금 더 디테일하게 말씀드려보겠습니다🙌
1번 항목에서 포인트로 제가 생각했던건 이러한 처리를 결국 우리는 print 디버깅으로 찍어볼건 아니고 추후 UI에서 어떤 재고가 부족한지 알아야하는게 크다고 생각해요🤔
우선 두괄식으로 말씀드려보면 에러 처리는 제일 말단 (그러니까 현재에서는 뷰컨이 되겠죠?)에서 처리하는게 맞다고 봅니다..!
그렇기에 결국 흐름을 생각해보면 UI에서 쥬스 주문 버튼을 누르면 해당 JuiceMaker 인스턴스의 orderFruitJuice 메서드를 호출할것이고 해당 메서드는 FruitStore 인스턴스의 chooseRecipe 메서드를 호출하는 로직일거라 생각되요.
그러면 chooseRecipe 메서드에서 재고가 부족하면 error를 해당 메서드를 부른 orderFruitJuice로 던질것이고 이해하신것과 마찬가지로 여기서 처리할 필요가 없고 해당 메서드는 throws 메서드가 되야할것이고 try로 chooseRecipe 메서드만 호출하면 될것 같습니다!
마지막으로 VC에서 orderFruitJuice 메서드를 부른쪽에서 do catch 처리를 해 얼럿으로 나타내주거나 해야될거라 생각합니다🙋🏻

p.s. 어후.. 쓰다보니 길어졌네요ㅠㅠ 결론은 이와 동일한 흐름으로 이해하셨다고 생각해 맞다는 소리입니다..!👍👍

apwierk2451 replied:
그린 금요일 5시가 넘어서 질문드려서 죄송했습니다 😅
VC에서 orderFuiltJuice 실행 및 오류처리 -> elert 로 코딩해보겠습니당~~
p.s. 감사합니다  좋은 주말 보내세요!

GREENOVER replied:
어후.. 아닙니다!!
5시 넘어서 질문해도 되요 정말입니다👍

GREENOVER reviewed (changes requested):
@ZZBAE @bradheo65 @apwierk2451 
안녕하세요 여러분!
PR로는 처음 인사드리네요🙌
우선 스텝1 산출한 기간에 아주 세부적으로 잘 구현해주셨습니다! 정말 정말 수고 많으셨씁니다👍👍
앞으로 3주간 쥬스메이커를 같이 잘 만들어봐요ㅎㅎ
잘 알고 계시겠지만 3주간 저는 여러분들과 한팀이 되어 스케쥴 관리에 대해서는 팀장의 역할처럼,
코드 리뷰에 대해서는 정말 동료 같은 레벨의 개발자로 의견을 나누려합니다!
이에 제 코멘트는 정말 참고사항이고 정답이 없기에 여러분들의 코드에 대한 더 뚜렷한 의견이 있다면 언제든 환영이고 저도 배워가겠습니다🙌
그럼 3주간 잘 부탁드립니다🙏🏻
***
### 칭찬 드리고 싶은 부분
- 서로 페어를 아주 잘 수행해주셨네요💯
- 모델 타입들이 잘 구분되어 있던점! 💯
- UML 클래스 다이어그램 작성이 처음이셨을것 같은데 잘 이해하고 작성해주신점 💯 (한눈에 이해하기 너무 좋았습니다!)
- 크.. 순서도를 그려보신것도 아주 최고네요 💯 (사실 순서도를 그려보는 캠퍼분들은 거의 오랜만에 보네요 최고👍)
- 명확한 이유로 프로토콜을 채택하고 잘 사용해주신점 💯
***
###  고민 및 해결한점에 대한 저의 의견
1. 열거형 프로토콜
- 따로 의견 첨가할 부분없이 CaseIterable를 정말 잘 이해하시고 명확하고 올바르게 써주셨던점 아주 좋았어요!
해당 allCases로 돌리는 부분의 코드도 고급지고 코드 가독도 매우 좋았습니다👍
2. for문과 forEach의 사용성
- 요 부분의 의견도 1번 항목의 의견과 아주아주 동일합니다!
연계되어 너무 잘해주셨고 명확한 이유로 이유있는 코드를 구현해주셨기에 최고라 생각합니다👍
3. try-catch 디폴트 오류
- 에러 처리에 대해 고민 잘해주시고 표현도 잘해주신것이 코드에 녹아있어서 이해가 잘 되었습니다..!
리뷰 코멘트로도 남기긴 했지만 요 부분에 대해 보완되어야 하는점이 있다면, 디폴트 에러에 대해 단순 error로 print 혹은 이후 UI에 얼럿으로 노출 시켜줘야한다면 아마 사용자에게는 불필요한 에러 정보일거라 생각해요🥲
이에 요 부분도 사용자의 입장에서 더 친숙하고 친절한 에러 처리가 있으면 어떨까 의견 드립니다🙋🏻
***
### 조언 요청에 대한 저의 의견
1. UML 작성 시 중요한 포인트
- 우선 UML 작성해주신것 아주 좋았고 베이스가 잘 되어있다고 생각합니다.
각 객체들이 채택하는 프로토콜도 명시를 잘 해주셨네요👍
여기서 더 발전될 수 있는 부분에 대해 포인트를 짚어볼까해요😀
우선 현재 클래스 다이어그램에서는 각 객체의 액션(메서드)들은 잘 정의 되어 있는데 해당 객체들이 가지는 프로퍼티에 대한 정의가 없기에 이 부분이 보완되면 좋을것 같습니다.
만드신 enum 타입의 다이어그램이 사실상 쓰이는 느낌으로 보이진 않아서 이 친구들을 연결시켜보면 좋겠다 싶었어요!
또 하나는 각 메서드들에서 필요한 파라미터 인자가 있으면 같이 표현해보면 더 명확할것 같습니다🙋🏻
이 부분들 외에 가장 중요한 포인트들(즉, 객체간 관계 및 모델 타입 정의)에 대해서 잘 해주셔서 10%정도 보완되면 완벽하겠습니다!
2. Struct VS Class
- 요 부분도 아주 잘 이해해주셨고 명확하게 이유있게 사용하셔서 더 조언드릴 부분은 없습니다!
저도 전적으로 동의합니다👍
***
여러분 다시 한번 스텝 1 수고 많으셨습니다..!
리뷰가 쪼오금.. 많은 느낌이 있긴하지만
전체적으로 한번 다시 체크해보고 다듬어보는 시간으로 생각해주시면 좋을것 같아요🙏🏻
리뷰 토대로 수정 혹은 의견을 달아주시면 좋겠습니다!
여러분의 명확한 의견이 있다면 수정하지 않아도 너무 좋습니다🙌
혹여 제가 잘못 이해했거나 이건 아닌거 같은데! 하는 부분이 있으면 언제든 디엠이나 코멘트나 편하게 말씀 부탁드려요~😄
언제든 리뷰 재요청을 해주시면 됩니다🙏🏻

bradheo65 committed:
style: 변수, 함수명 리뷰 참고사항에 따라 변경

ZZBAE committed:
style: 리뷰에 대한 변수와 함수 네이밍 변경

ZZBAE committed:
docs: Readme.md 생성

Step1 에 대한 내용 작성

apwierk2451 committed:
refact: FruitStore에 선언해 둔 함수 JuiceMaker로 옮기는 작업

apwierk2451 committed:
################

# <타입> : <제목> 의 형식으로 제목을 아래 공백줄에 작성
# 제목은 50자 이내 / 변경사항이 "무엇"인지 명확히 작성 / 끝에 마침표 금지
# 예) feat : 로그인 기능 추가
refact: FruitStore에 선언해 둔 함수 JuiceMaker로 옮기는 작업
함수의 매개변수 변경
# 바로 아래 공백은 지우지 마세요 (제목과 본문의 분리를 위함)

################
# 본문(구체적인 내용)을 아랫줄에 작성
# 여러 줄의 메시지를 작성할 땐 "-"로 구분 (한 줄은 72자 이내)

################
# 꼬릿말(footer)을 아랫줄에 작성 (현재 커밋과 관련된 이슈 번호 추가 등)
# 예) Close #7

################
# feat : 새로운 기능 추가
# fix : 버그 수정
# docs : 문서 수정
# test : 테스트 코드 추가
# refact : 코드 리팩토링
# style : 코드 의미에 영향을 주지 않는 변경사항
# chore : 빌드 부분 혹은 패키지 매니저 수정사항
################

bradheo65 committed:
Merge branch 'feature/step01' of https://github.com/ZZBAE/ios-juice-maker into feature/step01

GREENOVER commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -12,7 +12,15 @@ class ViewController: UIViewController {
-
+    var juiceMaker = JuiceMaker()
+    
+    func btnEvent(juice: Juice) {
```
추후 스텝에서 어떻게 쓰일지 기대됩니다👍
메서드명은 조금 더 명확히 변경될 여지가 있어 보입니다😄

GREENOVER commented on JuiceMaker/JuiceMaker/Error/StockError.swift:
```diff
@@ -0,0 +1,10 @@
+//
+
+enum StockError: Error {
+    case outOfStock
```
아직 이전 리뷰에서 반영안된 부분이 있어보이긴하네요..!
알 수 없는 에러들에 대해서도 한번 생각해보면 좋겠습니다🙋🏻

apwierk2451 replied:
case defaultError 를 만들려고 했는데 깜빡한 듯 하네요.. !🫣

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Fruit/Fruit.swift:
```diff
@@ -0,0 +1,29 @@
+    case kiwi
+    case mango
+    
+    var name: String {
+        switch self {
+        case .strawberry:
+            return "딸기"
+        case .banana:
+            return "바나나"
+        case .pineapple:
+            return "파인애플"
+        case .kiwi:
+            return "키위"
+        case .mango:
+            return "망고"
+        }
+    }
```
👍👍👍

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice/JuiceMaker.swift:
```diff
@@ -0,0 +1,40 @@
+
+struct JuiceMaker {
+    let fruitStore = FruitStore(defaultFruitStock: 10)
+    var insufficientStock: [Fruit:Int] = [:]
```
불충분한 재고? 느낌의 네이밍으로 보이는데 조금 불명확하지 않을까해요!
딕셔너리 타입으로 부족한 재고의 과일의 느낌을 한번에 줄 수 있는 네이밍은 어떤게 있을까요?🤔

GREENOVER replied:
추가로 보통 딕셔너리 선언 시 [Fruit: Int]와 같은 코드 컨벤션을 가집니다🙋🏻
다른 코드 부분에서는 이 컨벤션을 잘 지켜주셨던것 같아서 단순 깜빡하셨다고 보입니다🙃
https://developer.apple.com/documentation/swift/dictionary

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice/JuiceMaker.swift:
```diff
@@ -0,0 +1,40 @@
+        if insufficientStock.isEmpty {
+            useStock(juice: order)
+        } else {
+            throw StockError.outOfStock
```
현재의 outOfStock 정의로 어떤 재고들이 부족한지 에러 처리해주는 부분에서 파악할 수 있을까요~?

apwierk2451 replied:
insufficientStock 딕셔너리를 파라미터로 넘겨줄 예정입니다.

GREENOVER replied:
좋습니다! 그러면 outOfStock 케이스 시그니쳐도 바뀌어야되겠네요😀

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice/JuiceMaker.swift:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+    
+    private mutating func checkFruitStock(fruit: Fruit, amount: Int) throws {
+        guard let stockFruit = fruitStore.stocks[fruit] else { throw StockError.outOfStock }
+        guard stockFruit >= amount else {
+            return insufficientStock[fruit] = amount - stockFruit
+        }
+    }
```
이 메서드에서 에러를 throw해주신 의도는 어떻게 될까요?
어떤 로직 흐름을 생각하셨는지 궁금합니다..!
chooseRecipe 메서드에서 해당 checkFruitStock 메서드를 호출해서 부족한 재고에 대해서는 insufficientStock에 담아주고 결국 chooseRecipe 메서드에서 insufficientStock가 empty하지 않으면 "재고 부족" 에러를 던져주는걸로 이해가 되서요~!
만약 제가 잘못 이해하고 있는거라면 의견 말씀 부탁드립니다🙏🏻

apwierk2451 replied:
이 부분 case defaultError로 처리하겠습니다

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice/JuiceMaker.swift:
```diff
@@ -0,0 +1,40 @@
+    
+    private mutating func checkFruitStock(fruit: Fruit, amount: Int) throws {
+        guard let stockFruit = fruitStore.stocks[fruit] else { throw StockError.outOfStock }
+        guard stockFruit >= amount else {
+            return insufficientStock[fruit] = amount - stockFruit
+        }
```
guard else로 해주신 이유가 궁금합니다🙋🏻

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice/Juice.swift:
```diff
@@ -0,0 +1,35 @@
+    case kiwi
+    case pineapple
+
+    var recipeOfJuice: [Fruit: Int] {
```
간단히 recipe라는 네이밍은 어떨까요?
해당 프로퍼티로 접근하는 과정을 생각해보면 변수를 따로 지정해주지 않았다면 Juice.recipeOfJuice가 될것입니다.
시그니쳐대로 읽어보면 `쥬스쥬스의재료`가 되어 Juice라는게 중복되어 어색하게 보입니다🤔
이 Juice라는 친구를 현재 chooseRecipe에서 파라미터 인자로 order로 받고 계신것은 확인이 됐고 그 부분만 보자면 order recipeOfJuice라는건 자연스럽고 어색하지 않습니다.
다만 모든 타입들을 독단적으로 봤을때 어색하지 않는것이 좋음으로 해당 프로퍼티 네이밍을 변경했을때 order와의 쓰임에서 어색해지는 부분이 있다면 order 파라미터 인자 네이밍을 변경하는쪽이 더 맞지 않을까 싶습니다🙋🏻

GREENOVER commented on README.md:
```diff
@@ -0,0 +1,266 @@
+# ios-juice-maker
```
크... 리드미 작성 최고네요👍👍👍👍
정말 분명히 꾸준한 리드미 작성이 여러분의 무기로 돌아올거에요!!

GREENOVER commented on JuiceMaker/JuiceMaker.xcodeproj/project.pbxproj:
```diff
@@ -41,6 +47,32 @@
+			path = Fruit;
+			sourceTree = "<group>";
+		};
+		A3114A35281968A600C903B1 /* Error */ = {
+			isa = PBXGroup;
+			children = (
+				18A346C528181E28003F2F29 /* StockError.swift */,
+			);
+			path = Error;
+			sourceTree = "<group>";
+		};
```
👍👍👍👍👍

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice/JuiceMaker.swift:
```diff
@@ -0,0 +1,40 @@
+            try checkFruitStock(fruit: fruit, amount: fruitAmount)
+        }
+        if insufficientStock.isEmpty {
+            useStock(juice: order)
```
의견차이일 수 있는데 요 재고를 사용해주는 메서드가 makeJuice쪽에 와야 되지 않을까 생각했습니다!
메서드명만 봐서 관심사가 다른 느낌이 들어가지구요..!
재고를 사용하는것은 쥬스를 만드는쪽에 가깝지 레시피를 고르는쪽과는 약간 역할이 다르지 않나 생각이 듭니다🤔

GREENOVER commented on JuiceMaker/JuiceMaker/Model/Juice/JuiceMaker.swift:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+    
+    private mutating func checkFruitStock(fruit: Fruit, amount: Int) throws {
+        guard let stockFruit = fruitStore.stocks[fruit] else { throw StockError.outOfStock }
+        guard stockFruit >= amount else {
+            return insufficientStock[fruit] = amount - stockFruit
+        }
+    }
+    
+    private func useStock(juice: Juice) {
```
checkFruitStock과 useStock의 구현부는 JuiceMaker가 가지고 있는게 맞을까 싶습니다!
결국 재고를 관리해주고 체크해주고 하는 역할은 FruitStore에서 해줘야하지 않을까해요🤔

GREENOVER reviewed (approved):
세분 모두 리팩토링 해주시느라 고생 많으셨습니다👍👍👍
훨씬 더 명확하고 보완되었어요!
특히 리드미 추가 아주 나이스 합니다😄😁
몇가지 보완 혹은 의견을 여쭤보는 코멘트들을 남겼습니다..!
이후 스텝으로 가기에는 전혀 문제되는 부분이 아님으로 해당 PR은 어프로브 후 머지하겠습니다🙋🏻
남겨진 코멘트에 대해서는 다음 스텝을 진행하면서 보완 혹은 생각해보시면 될것 같아요~!
스텝 1 너무 수고하셨고 다음 스텝도 화이팅입니다🥳🥳🥳

p.s. 아..! 이후에 리뷰 재요청을 해주실때 디코도 좋지만 깃헙에 요런 기능이 있습니다.
PR 우측 상단에 저 빨간 박스를 눌러주시면 리뷰 재요청을 저에게 날려줍니다ㅎㅎ
요걸 활용하시면 제가 깃헙을 보고 있다면 알림이 와서 더 빠르게 캐치할 수 있습니다!
물론 저것도 해주시고 디코로도 디엠 주시는게 제일 베스트라ㅎㅎ.. 제가 빠르게 리뷰 드릴 수 있도록 둘다 해주시면 감사하겠습니다
![스크린샷 2022-05-01 오전 11 11 32](https://user-images.githubusercontent.com/72292617/166130289-1003fe5d-f855-4661-ac1c-6ae3c89a7b29.png)
